Remove selector stuff from VarInfo tests and link/invlink

mhauru opened the pull request:
This PR implements a part of moving away from using Selectors/Gibbs IDs/indexing by samplers. It does two things:
* It removes any and all mentions of indexing `VarInfo`s by samplers in `test/varinfo.jl`.
* It reworks `link`, `link!!`, `invlink`, and `invlink!!` to not take samplers as arguments anymore, but rather (iterables of) `VarName`s.

These two changes were easiest to do at the same time, for reasons.

Note that more clean up Selector/Gibbs ID stuff is to come, but I'm asking for code review at this stage to avoid one megareview. This PR is not to be merged to `master`, but rather to #779, which will collect all Selector-removal changes.

mhauru commented:
Okay, we are back.

sunxd3 commented:
Markus, could you say something about why using sampler for `link` is not a good idea? (I do prefer the current approach, but just to be more certain). Particularly the logic/reason behind using sampler to index `link` (before) and using varnames to index `link` (after)?

mhauru commented:
Sure, @sunxd3. This all part of a general push to get rid of marking different variables as belonging to different samplers in `VarInfo` (#573). By the time #779 is ready to merge, I expect Ctrl+F of "Sampler" in `varinfo.jl` to bring up nothing.

I view this mostly as a modularity/separation of concerns thing: VarInfo's job is to keep track of which variable has which value, and how they've been transformed, and maybe what the accumulated logprob is. What exactly is being done the model that warrants keeping track of these things, like whether we are drawing a single sample from the prior or doing MCMC or variational inference or predicting or whatever it is, is a higher level issue downstream of VarInfo, and VarInfo should be blind to it.

To be philosophical about it, this is how I think it currently looks:
```
                                                                             
 ┌──────────────────────────────────────────┬──────────────────────────────┐ 
 │                                          │                              │ 
 │          ┌───────┐               VarInfo │ Sampler                      │ 
 │          └─────┐ │                       │                              │ 
 │ ┌─────┐        │ └──────────┐            │                              │ 
 │ └───┐ │        │   ┌─────┐  │            │                              │ 
 │     │ └────────┘   └─┐   └─┐└────┐       │                              │ 
 │     └──────────────┐ │     └───┐ │       │                              │ 
 │                    │ │         │ └───────┘                              │ 
 │                    │ │         └─────────┐                              │ 
 │                    │ │ ◄──────┐    ▲     │                              │ 
 │                    │ │        ├────┘     │                              │ 
 │      ┌─────────────┘ │        │          │                              │ 
 │      │ ┌─────────────┘        │          │                              │ 
 │      │ │                      │          │                              │ 
 │      │ │     Indexing a VarInfo by       │                              │ 
 │      └─┘     samplers                    │                              │ 
 └──────────────────────────────────────────┴──────────────────────────────┘ 
```

and this is how I think it should look:

```
 ┌───────────────────────────┬──────────────────────────────┐ 
 │                           │                              │ 
 │                   VarInfo └──┐ Sampler                   │ 
 │                              │                           │ 
 │                           ┌──┘                           │ 
 │                           │                              │ 
 │                           └──┐                           │ 
 │                              │                           │ 
 │                           ┌──┘                           │ 
 │                           │                              │ 
 │                           └──┐                           │ 
 │                              │                           │ 
 │                           ┌──┘                           │ 
 │                           │                              │ 
 └───────────────────────────┴──────────────────────────────┘ 
```

There are also some more practical benefits:
* The `VarInfo` code can be greatly simplified, and we can switch over to using `VarNamedVector`. Probably some edge case bugs will be fixed as a side effect too.
* We can have a well-defined interface for `VarInfo`, that is stable and properly documented. This makes code easier to reason about.
* Turing's own samplers can be on equal footing with external samplers. Currently a sampler needs to be a of a particular type to be used with `VarInfo`, and conforming to that requires glue code in Turing.jl. In the future a sampler can just take a state of the sampling process, as a vector or maybe as a `VarInfo`, modify it and return it, using a natural interface where it just sees a bunch of variables with values, and is blind to the internals of how `VarInfo` works.

Note that the only reason (I think) why this indexing of `VarInfo`s by samplers was introduced in the first place was Gibbs sampling. The old Gibbs sampler used this mechanism to keep track of which component sampler applied to which variable. Now with the new Gibbs sampler that is no longer needed.

sunxd3 commented:
Thanks a lot for the explanation, crystal clear!

sunxd3 commented on src/abstract_varinfo.jl:
```diff
@@ -548,37 +550,67 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
-    return link!!(t, vi, SampleFromPrior(), model)
+    vns = collect(keys(vi))
+    # In case e.g. vns = Any[].
+    # TODO(mhauru) Could we rather fix akeys` so that it would always return VarName[]?
```
```suggestion
    # TODO(mhauru) Could we rather fix `keys` so that it would always return VarName[]?
```
typo?

sunxd3 commented on src/abstract_varinfo.jl:
```diff
@@ -548,37 +550,67 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
 See also: [`default_transformation`](@ref), [`invlink!!`](@ref).
 """
-link!!(vi::AbstractVarInfo, model::Model) = link!!(vi, SampleFromPrior(), model)
+# Use `default_transformation` to decide which transformation to use if none is specified.
```
Julia will think a docstring belongs to a function only when the function is right next to the docstring (second paragraph of [this](https://docs.julialang.org/en/v1/manual/documentation/#Writing-Documentation)). Kind of annoying, but

penelopeysm replied:
@sunxd3 I get the sense that you got caught out by this once and now you've taken it upon yourself to warn all of us 😄

sunxd3 commented on src/abstract_varinfo.jl:
```diff
@@ -548,37 +550,67 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
 See also: [`default_transformation`](@ref), [`invlink`](@ref).
 """
-link(vi::AbstractVarInfo, model::Model) = link(vi, SampleFromPrior(), model)
+# Use `default_transformation` to decide which transformation to use if none is specified.
```
this will also cause doc issue

mhauru replied:
Done

sunxd3 commented on src/abstract_varinfo.jl:
```diff
@@ -587,49 +619,81 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
 See also: [`default_transformation`](@ref), [`link!!`](@ref).
 """
-invlink!!(vi::AbstractVarInfo, model::Model) = invlink!!(vi, SampleFromPrior(), model)
+# Use `default_transformation` to decide which transformation to use if none is specified.
```
here

mhauru replied:
Done

sunxd3 commented on src/abstract_varinfo.jl:
```diff
@@ -587,49 +619,81 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
+    # loudly.
+    all_vns = Set(keys(vi))
+    if Set(vns) != all_vns
+        msg = "StaticTransforming only a subset of variables is not supported."
```
```suggestion
        msg = "Statically transforming only a subset of variables is not supported."
```
maybe?

sunxd3 commented on src/abstract_varinfo.jl:
```diff
@@ -638,12 +702,25 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
 See also: [`default_transformation`](@ref), [`link`](@ref).
 """
-invlink(vi::AbstractVarInfo, model::Model) = invlink(vi, SampleFromPrior(), model)
+# Use `default_transformation` to decide which transformation to use if none is specified.
```
here

mhauru replied:
Done

sunxd3 reviewed (approved):
Solid improvements, thanks.

penelopeysm commented on src/utils.jl:
```diff
@@ -1268,3 +1271,65 @@ _merge(left::NamedTuple, right::NamedTuple) = merge(left, right)
+    return NamedTuple{(T,)}((vns,))
+end
+
+varname_namedtuple(vns::NamedTuple) = vns
```
Is this method needed? I don't think it's used anywhere?

I ask because the other implementations provide some sort of guarantee that the returned NamedTuple obeys certain properties (i.e. the keys are the unique symbols, the values are grouped...) but this one doesn't.

mhauru replied:
Not needed any more, removed.

penelopeysm commented on src/varinfo.jl:
```diff
@@ -1183,24 +1212,41 @@ function _link!(vi::UntypedVarInfo, spl::AbstractSampler)
+function _link!(vi::TypedVarInfo, vns::VarNameCollection)
+    return _link!(vi, varname_namedtuple(vns))
 end
-function _link!(vi::TypedVarInfo, spl::AbstractSampler, spaceval::Val)
-    vns = _getvns(vi, spl)
-    return _link!(vi.metadata, vi, vns, spaceval)
+
+function _link!(vi::TypedVarInfo, vns::NamedTuple)
+    return _link!(vi.metadata, vi, vns)
+end
```
I suppose that `varname_namedtuple(::NamedTuple)` would become relevant if you merged these two methods into one.

mhauru replied:
Yeah, that's the reason `varname_namedtuple(::NamedTuple)` was defined. I've now removed it. These two needed to be separated to method disambiguate.

penelopeysm reviewed (changes requested):
My overarching feeling about this so far is that we are being really loose with types. I think it's great to offer the user some convenience, but in this case I fear that we have too many pathways which makes it very hard to reason about. e.g. if you pass a tuple to `link!!`, it's really hard to keep track of the output type, and then that output type becomes the input to another function which might convert it to a namedtuple, but in a type stable way; but on the other hand if you pass it a vector, sure, the eventual outcome is the same, but it follows a rather different winding path through `abstract_varinfo.jl` and `varinfo.jl`.

The main question I have is really whether we need this flexibility? All the linking functions are entirely internal to Turing and so external user convenience isn't really a consideration.

If we made sure that all of these functions could only take one container, then it appears to me that we could potentially:

(1) cut some of the code that is repeated
(2) guarantee that it proceeds via a path that is type stable
(3) make it easier to follow the dispatch chain when we revisit this

which would be a win in my books 😄 and I'd be super, super happy to wrap my lists in a tuple before calling `link!!` if it means that we get the above.

penelopeysm reviewed (commented):
Just a handful more comments to the previous

torfjelde commented:
This is great @mhauru :tada:

I'm a bit OOL, but my recommendation would be *against* exposing varname-specific linking to the "user".

As you said, ideally we'd move away from this (and it can be done by a combination of `subset` and `merge`), and it will simplify things quite a bit I'd imagine.

I would do:

1.  Define `link` / `invlink` for the entire `VarInfo`, which simply defers to its `metadata`.
2.  Define `link` / `invlink` only for `(metadata, varname)` pairs, i.e. no tuples or vectors.

Is there a reason why this wouldn't work / is not ideal?

Happy to have a quick chat about all of this sometime this week if you want:)

torfjelde commented:
Also loved the illustrations above :sweat_smile:

mhauru commented on src/abstract_varinfo.jl:
```diff
@@ -548,37 +550,65 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
-    return link(t, deepcopy(vi), SampleFromPrior(), model)
+    vns = collect(keys(vi))
+    # In case e.g. vns = Any[].
+    # TODO(mhauru) Could we rather fix akeys` so that it would always return VarName[]?
```
Issue here: https://github.com/TuringLang/DynamicPPL.jl/issues/791

mhauru commented on src/abstract_varinfo.jl:
```diff
@@ -548,37 +550,65 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
+function link!!(vi::AbstractVarInfo, model::Model)
+    return link!!(default_transformation(model, vi), vi, model)
+end
+function link!!(vi::AbstractVarInfo, vns, model::Model)
```
This is actually significant, because underspecifying the type avoids a method ambiguity. That in itself is a clear sign that the multiple dispatch use here is too hard to understand. I won't fix this before we've decided what to do about the Vector methods though.

mhauru replied:
Fixed.

mhauru commented on src/utils.jl:
```diff
@@ -1268,3 +1271,65 @@ _merge(left::NamedTuple, right::NamedTuple) = merge(left, right)
+`Base.unique`. The point of `unique_syms` is that it supports constant propagating
+the result, which is possible only when the argument and the return value are `Tuple`s.
+"""
+@generated function unique_syms(::T) where {T<:NTuple{N,VarName}} where {N}
+    retval = Expr(:tuple)
+    syms = [first(vn.parameters) for vn in T.parameters]
+    for sym in unique(syms)
+        push!(retval.args, QuoteNode(sym))
+    end
+    return retval
+end
```
I assume this comment is kinda superseded by the wider conversation about whether to keep the vector methods. Note that `unique_syms` is important for the type stability of the tuple methods.

mhauru replied:
We no longer support vectors, so this is sure to be called when calling `link/invlink(::TypedVarInfo, ::NTuple{N,VarName} where {N}, model)`. On that path this is significant for type stability, and thus I presume for performance. Let me know if you are happy with the current situation.

penelopeysm replied:
Yes, I fully approve of that :)

mhauru commented on src/transforming.jl:
```diff
@@ -91,14 +91,16 @@ function dot_tilde_assume(
     return r, lp, vi
 end
 
+VarNameCollection = Union{NTuple{N,VarName} where N,AbstractVector{<:VarName},NamedTuple}
```
Done

mhauru commented on src/varinfo.jl:
```diff
@@ -1966,50 +2053,49 @@ function unset_flag!(vnv::VarNamedVector, ::VarName, flag::String, ignorable::Bo
 end
-function set_retained_vns_del_by_spl!(vi::TypedVarInfo, spl::Sampler)
+function set_retained_vns_del!(vi::TypedVarInfo)
     # Get the indices of `vns` that belong to `spl` as a NamedTuple, one entry for each symbol
```
Done

mhauru commented on test/varinfo.jl:
```diff
@@ -856,73 +823,6 @@ end
-            @test DynamicPPL.getgid(varinfo_merged, @varname(y)) == gidset_right
-        end
-
-        # The below used to error, testing to avoid regression.
-        @testset "merge different dimensions" begin
-            vn = @varname(x)
-            vi_single = VarInfo()
-            vi_single = push!!(vi_single, vn, 1.0, Normal())
-            vi_double = VarInfo()
-            vi_double = push!!(vi_double, vn, [0.5, 0.6], Dirichlet(2, 1.0))
-            @test merge(vi_single, vi_double)[vn] == [0.5, 0.6]
-            @test merge(vi_double, vi_single)[vn] == 1.0
-        end
```
Did not intend to, thanks for catching. Restored.

mhauru commented on src/abstract_varinfo.jl:
```diff
@@ -539,7 +539,9 @@ function settrans!! end
+    link!!([t::AbstractTransformation, ]vi::AbstractVarInfo, vn::VarName, model::Model)
+    link!!([t::AbstractTransformation, ]vi::AbstractVarInfo, vn::Tuple{N,VarName}, model::Model)
+    link!!([t::AbstractTransformation, ]vi::AbstractVarInfo, vn::AbstractVector{<:VarName}, model::Model)
 
```
I've removed this now, even though I would like to have the single-VarName version. The reason I've removed it is that the dispatch on `link`/`invlink` is already too complicated, which indeed causes method ambiguities. I think if the dispatch got cleaned up otherwise we could reintroduce the single-VarName version, which we should be able to do with just a single method without ambiguities (because no other method would accept a single VarName). However, for now it's not worth the code complexity, given that it's not really needed anywhere.

For the first point of being able to see the type at the call site, I see what you mean, but I think this is something you have to largely give up in a dynamically typed language like Julia, and I think the ergonomy of the single-VarName method would outweigh it.

penelopeysm replied:
> I think this is something you have to largely give up in a dynamically typed language like Julia

Maybe not very surprising, but I don't agree with this 🙈 

Dynamic typing means that it's not possible to unambiguously determine the type of something (using type inference algorithms, or in practice, LSP), but often it's possible to make limited inferences from the known function signatures. In a dynamically typed language, it's not possible to verify these inferences at compile time, but IMO it doesn't follow from this that we should make it harder for ourselves to make such inferences.

The most frequently cited benefit of dynamic typing is the ease of prototyping. However, I think this primarily applies to the end users of the language, i.e. people who end up using Turing. If this was a case of providing nice APIs for end users, I'd be totally on board with extra convenience methods, but I think the `(inv)?link(!!)?` functions are pretty much entirely for internal use. I believe (and I hope it's not a controversial opinion!) that it's easier to maintain, reason about, and guarantee the correctness of software that is statically typed, and although we can't do that in Julia, I do think our lives will be easier if we are more disciplined about our internal APIs 😉 

Anyway, I just wanted to write it out because I suspect that me pushing back against multiple dispatch 'overuse' (quotes because others may not agree that it's overuse!) will be a continuing trend in PRs. I do recognise it's not my personal codebase and I'll always be happy to compromise on the outcome, in the sense that if you say you really want a convenience method I'll just say 'ehh, whatever', but I think it's unlikely that I'll really change my mind on the underlying principle. So I write it out once and I'll spare all of you this next time 😄

penelopeysm replied:
(Btw just on this PR I'm happy with how it stands with dispatch so feel free to resolve this convo!)

mhauru commented on src/abstract_varinfo.jl:
```diff
@@ -548,37 +550,65 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
+function link!!(vi::AbstractVarInfo, vns, model::Model)
+    return link!!(default_transformation(model, vi), vi, vns, model)
+end
+# If no variable names are provided, link all variables.
```
Fixed. Would still like to understand if this would have caused issues with docstrings.

penelopeysm replied:
Oh, I'm wrong on this one: it is just the function, not the method.

mhauru commented on src/abstract_varinfo.jl:
```diff
@@ -548,37 +550,65 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
-function link!!(vi::AbstractVarInfo, spl::AbstractSampler, model::Model)
-    # Use `default_transformation` to decide which transformation to use if none is specified.
-    return link!!(default_transformation(model, vi), vi, spl, model)
+# Wrap a single VarName in a singleton tuple.
```
Done

mhauru commented on src/abstract_varinfo.jl:
```diff
@@ -548,37 +550,65 @@ If `t` is not provided, `default_transformation(model, vi)` will be used.
-function link(vi::AbstractVarInfo, spl::AbstractSampler, model::Model)
-    # Use `default_transformation` to decide which transformation to use if none is specified.
-    return link(default_transformation(model, vi), deepcopy(vi), spl, model)
+# Wrap a single VarName in a singleton tuple.
```
Done all

mhauru commented:
I searched in TuringLang for all uses of `link`/`invlink`, and internally we don't use anything other than the `link/invlink(varinfo, model)` version, which links all variables. Hence I'm happy with not having the vector option, and I've removed it.

I also removed the single `VarName` version just to simplify dispatch and reduce the multitude of methods. I would ideally like to reintroduce it at some point since I think it's nice for ergonomics, but since we don't need it I think the simplicity is worth it for now.

I would like to keep some way for users to manually link individual variables, so I've left the tuple version. I would like to keep it because we export `merge`, which allows a user to create `VarInfo`s with mixed linking, and thus I think they should have the ability to manipulate the linking as well.

In the process I also added some code duplication to reduce the use of `Union` types, make the dispatch cascade easier to reason about, and make some error messages more informative. The principle is now that for each subtype of `AbstractVarInfo` you have to separately define `link` and `invlink` for the set of arguments that includes a tuple of `VarName`s and for the set that doesn't; i.e. there is no fallback where if no list of `VarName`s is given we would get a default option with `keys(varinfo)`.

penelopeysm commented on src/abstract_varinfo.jl:
```diff
@@ -537,113 +537,116 @@ If `vn` is not specified, then `istrans(vi)` evaluates to `true` for all variabl
 """
     link!!([t::AbstractTransformation, ]vi::AbstractVarInfo, model::Model)
-    link!!([t::AbstractTransformation, ]vi::AbstractVarInfo, spl::AbstractSampler, model::Model)
+    link!!([t::AbstractTransformation, ]vi::AbstractVarInfo, model::Model)
+    link!!([t::AbstractTransformation, ]vi::AbstractVarInfo, vns::Tuple{N,VarName}, model::Model)
```
```suggestion
    link!!([t::AbstractTransformation, ]vi::AbstractVarInfo, vns::NTuple{N,VarName}, model::Model)
```
(The first line is duplicated)

penelopeysm commented on src/abstract_varinfo.jl:
```diff
@@ -537,113 +537,116 @@ If `vn` is not specified, then `istrans(vi)` evaluates to `true` for all variabl
 """
     link([t::AbstractTransformation, ]vi::AbstractVarInfo, model::Model)
-    link([t::AbstractTransformation, ]vi::AbstractVarInfo, spl::AbstractSampler, model::Model)
+    link([t::AbstractTransformation, ]vi::AbstractVarInfo, model::Model)
+    link([t::AbstractTransformation, ]vi::AbstractVarInfo, vns::Tuple{N,VarName}, model::Model)
```
```suggestion
    link([t::AbstractTransformation, ]vi::AbstractVarInfo, vns::NTuple{N,VarName}, model::Model)
```

penelopeysm commented on src/abstract_varinfo.jl:
```diff
@@ -537,113 +537,116 @@ If `vn` is not specified, then `istrans(vi)` evaluates to `true` for all variabl
 """
     invlink!!([t::AbstractTransformation, ]vi::AbstractVarInfo, model::Model)
-    invlink!!([t::AbstractTransformation, ]vi::AbstractVarInfo, spl::AbstractSampler, model::Model)
+    invlink!!([t::AbstractTransformation, ]vi::AbstractVarInfo, vns::Tuple{N,VarName}, model::Model)
```
```suggestion
    invlink!!([t::AbstractTransformation, ]vi::AbstractVarInfo, vns::NTuple{N,VarName}, model::Model)
```

penelopeysm commented on src/abstract_varinfo.jl:
```diff
@@ -537,113 +537,116 @@ If `vn` is not specified, then `istrans(vi)` evaluates to `true` for all variabl
 """
     invlink([t::AbstractTransformation, ]vi::AbstractVarInfo, model::Model)
-    invlink([t::AbstractTransformation, ]vi::AbstractVarInfo, spl::AbstractSampler, model::Model)
+    invlink([t::AbstractTransformation, ]vi::AbstractVarInfo, vns::Tuple{N,VarName}, model::Model)
```
```suggestion
    invlink([t::AbstractTransformation, ]vi::AbstractVarInfo, vns::NTuple{N,VarName}, model::Model)
```

penelopeysm commented on src/utils.jl:
```diff
@@ -1268,3 +1271,63 @@ _merge(left::NamedTuple, right::NamedTuple) = merge(left, right)
+Return the unique symbols of the variables in `vns`.
+
+Note that `unique_syms` is only defined for `Tuple`s of `VarName`s and, unlike
+`Base.unique`, returns a `Tuple`. For an `AbstractVector{<:VarName}` you can use
+`Base.unique`. The point of `unique_syms` is that it supports constant propagating
```
Maybe the line about AbstractVector isn't needed anymore?

mhauru replied:
Removed

penelopeysm commented on src/utils.jl:
```diff
@@ -1268,3 +1271,63 @@ _merge(left::NamedTuple, right::NamedTuple) = merge(left, right)
+`Base.unique`. The point of `unique_syms` is that it supports constant propagating
+the result, which is possible only when the argument and the return value are `Tuple`s.
+"""
+@generated function unique_syms(::T) where {T<:NTuple{N,VarName}} where {N}
```
```suggestion
@generated function unique_syms(::T) where {T<:NTuple{N,VarName},N}
```
You could remove one `where` here

mhauru replied:
Changed it to just say `T<:VarNameTuple`.

penelopeysm commented on src/utils.jl:
```diff
@@ -1268,3 +1271,63 @@ _merge(left::NamedTuple, right::NamedTuple) = merge(left, right)
+    return NamedTuple{syms}(elements)
+end
+
+# This method is type unstable, but that can't be helped: The problem is inherently type
+# unstable if there are VarNames with multiple symbols in a Vector.
+function varname_namedtuple(vns::AbstractVector{<:VarName})
+    syms = tuple(unique(map(getsym, vns))...)
+    elements = tuple((filter(vn -> getsym(vn) == s, vns) for s in syms)...)
+    return NamedTuple{syms}(elements)
+end
+
+# A simpler, type stable implementation when all the VarNames in a Vector have the same
+# symbol.
+function varname_namedtuple(vns::AbstractVector{<:VarName{T}}) where {T}
+    return NamedTuple{(T,)}((vns,))
+end
```
These methods maybe aren't needed any more?

mhauru replied:
Removed

penelopeysm commented on src/utils.jl:
```diff
@@ -1268,3 +1271,63 @@ _merge(left::NamedTuple, right::NamedTuple) = merge(left, right)
+    return retval
+end
+
+"""
+    varname_namedtuple(vns::NTuple{N,VarName}) where {N}
+    varname_namedtuple(vns::AbstractVector{<:VarName})
+    varname_namedtuple(vns::NamedTuple)
+
+Return a `NamedTuple` of the variables in `vns` grouped by symbol.
+
+`varname_namedtuple` is type stable for inputs that are `Tuple`s, and for vectors when all
+`VarName`s in the vector have the same symbol. For a `NamedTuple` it's a no-op.
+
```
Sorry to bikeshed, but would you be open to calling this `group_varnames_by_symbol`? I think it's more descriptive.

mhauru replied:
Much better, changed.

penelopeysm commented on src/varinfo.jl:
```diff
@@ -897,6 +900,21 @@ end
     return :($(exprs...),)
 end
 
+"""
+    all_varnames_namedtuple(vi::TypedVarInfo)
+
+Return a `NamedTuple` of the variables in `vi` grouped by symbol.
+"""
+all_varnames_namedtuple(vi::TypedVarInfo) = all_varnames_namedtuple(vi.metadata)
+
+@generated function all_varnames_namedtuple(md::NamedTuple{names}) where {names}
```
(If you accepted the above: this one could be `all_varnames_grouped_by_symbol` too)

penelopeysm commented on src/varinfo.jl:
```diff
@@ -1320,59 +1403,72 @@ function _getvns_link(varinfo::TypedVarInfo, spl::SampleFromPrior)
 end
 
-function _link(model::Model, varinfo::TypedVarInfo, spl::AbstractSampler)
+# If we try to _invlink! a TypedVarInfo with a Tuple or Vector of VarNames, first convert
+# it to a NamedTuple that matches the structure of the TypedVarInfo.
```
```suggestion
# If we try to `link!!` a TypedVarInfo with a Tuple of VarNames, first convert it to a
# NamedTuple that matches the structure of the TypedVarInfo.
```

mhauru replied:
I fixed the Vector thing and line length, but made the name say `_link`, since that's the function this is commenting on.

penelopeysm replied:
Ah, yes, I was just looking at it again and I think I misinterpreted the intent of the comment. I thought it was referring to the function that called it - I see it now. So feel free to ignore the others too

penelopeysm commented on src/varinfo.jl:
```diff
@@ -1444,57 +1540,76 @@ function _link_metadata!!(
 end
 
-function _invlink(model::Model, varinfo::TypedVarInfo, spl::AbstractSampler)
+# If we try to _invlink a TypedVarInfo with a Tuple or Vector of VarNames, first convert
+# it to a NamedTuple that matches the structure of the TypedVarInfo.
```
```suggestion
# If we try to `invlink!!` a TypedVarInfo with a Tuple of VarNames, first convert it to a
# NamedTuple that matches the structure of the TypedVarInfo.
```

mhauru replied:
I fixed the Vector thing and line length, but left the name as `_invlink`, since that's the function this is commenting on.

penelopeysm reviewed (commented):
Just some minor stuff remaining

mhauru commented on src/abstract_varinfo.jl:
```diff
@@ -715,9 +788,10 @@ function maybe_invlink_before_eval!!(
+    t::StaticTransformation, vi::AbstractVarInfo, ::AbstractContext, model::Model
 )
-    return invlink!!(t, vi, _default_sampler(context), model)
+    # TODO(mhauru) Why does this function need the context argument?
+    return invlink!!(t, vi, model)
 end
 
 function _default_sampler(context::AbstractContext)
```
Nice, thanks for doing that research, removed both.

penelopeysm commented:
I'll go for a walk, if I come back and CI has passed i'll approve 😄 

Last question, which doesn't necessarily need to be acted upon now: I wonder if it's worth keeping a note of what was changed somewhere. DPPL doesn't have a changelog (not yet, at least, my other PR puts one in) but it makes sense to me that each PR should add / modify the changelog entry so that when we do the big 0.35 merge we're not trying to collate all the changes. 
I figure that this might be an intermediate stage, so maybe you don't want to write something now, but I thought it might just be something to keep in mind

penelopeysm commented:
Thanks for spending the time to clean it up _and_ to bear with all my comments 😄

penelopeysm commented:
Go for it!
